ENH: special: Overhaul of Legendre functions

izaid opened the pull request:
This PR demonstrates a way forward for special functions defined by recurrence relations, as well as some other issues with special functions more generally. Here we will use the Legendre polynomials `lpn`, `lpmn`, and `clpmn` as illustrative examples, and the work has been done for these functions.

As has been discussed extensively elsewhere, a good model for special functions is C / C++ kernels published to Python as NumPy universal functions [1]. Even the more complicated functions that need to operate on subarrays, like the Legendre polynomials being discussed here, fit into this model as generalized universal functions [2]. By refactoring the existing special functions to be ufuncs, we are also able to provide the scalar kernels to other array libraries, so our model is more general than just NumPy on the CPU.

An unsolved question so far is then how to handle _variants_ of the normal behaviour. This is necessary for some of the more complicated special functions and hasn't been properly solved before. Consider the associated Legendre polynomial `lpmn` for instance. It should be able to support:

- Returning either the normalized or unnormalized associated Legendre polynomials. Currently only the unnormalized associated Legendre polynomials are available in SciPy, but everyone else (Julia, MATLAB, ...) supports normalization. Why does SciPy not?
- Returning derivatives, also up to different orders. Currently `lpmn` has to return both the polynomials `Pmn_z` and their derivatives `Pmn_d_z`. There is no option to not return the derivative. And throughout the special functions there are many different adhoc conventions for keywords that return derivatives.
- Returning all the values computed throughout the recurrence. These pairs of "at" versus "all" functions have already been discussed a lot [4].

Here I have refactored `lpn`, `lpmn`, and `clpmn` in both C++ and Python so that all these things can be supported easily. The C++ code is important but, for the purposes of SciPy, it is a private implementation. In Python, there are a couple of changes to the public SciPy API in the `special` module. I will describe everything below.

_Python API changes._
- `lpn`, `lpmn`, and `clpmn` all gain a keyword `diff_n` that indicates the order of derivatives to return. It defaults to 0, but can be 1 or 2. All derivatives up to that order are returned, so `diff_n = 2` gives you 3 outputs. This is just what makes sense for Legendre polynomials. Note I am happy to change the name of this keyword if suggested.
- The "all" versions of `lpn`, `lpmn`, and `clpmn` are introduced as `lpn_all`, `lpmn_all`, and `clpmn_all`. I thought long and hard about whether they should be in like that or as a keyword similar to `mode = "at"` or `mode = "all"`. Given how things are implemented, it just doesn't make a lot of sense to use a keyword here, so I've gone with separate functions.
- `lpn`, `lpmn`, and `clpmn` historically have returned the "all" versions, but will become the "at" versions going forward. There is a keyword `legacy` that defaults to `True` to switch the behaviour. There will need to be a deprecation cycle.
- `lpmn` and `clpmn` also gain a keyword `norm` which can be either `True` or `False` to return the normalised or unnormalised Legendre polynomials. I still have to implement the normalisation, but that will come once everyone is happy with all of this.

_C++ implementation_
- To support all these different variants (derivative or not, "all" or "at", and so on), I have rewritten the C++ algorithms to take a C++ callback. This callback is called at each recurrence step and can be fully custom. I've added callback functions to work out the derivatives and callbacks to store all the recurrence values in a subarray, this is how we can support all these different behaviours without writing the same algorithm many times. It should be effectively zero cost as C++ allows templating on the callback type, which is what I have done. The `lpn` C++ code is a clear example of how this works, see https://github.com/scipy/scipy/blob/79421615dc66425bc47899f2afe254f5c7284398/scipy/special/special/legendre.h#L22.
- Not all behaviour can be added on like that, and normalisation is a good case where it is just a different function.
- `clpmn` and `lpmn` have been merged into one function, but defaulting to different branch cuts just like SciPy currently does.
- Each pair of `diff_n` and `norm` is published to Python as a ufunc. In this case, they are collected using `Py_BuildValue` as a dictionary of tuples. So essentially what we have is a simple and minimal collection of ufuncs that encapsulates all the behaviour variants.

_Python implementation_
- At the Python level, I have introduced a minimal class called `multiufunc`. This is only intended to be used for these special functions with complicated behaviour, and will behave like a ufunc as much as possible. It has a very simple model.
- In a `multiufunc`, keywords can be used to select the right ufunc from the collection of ufuncs. So `lpmn(m, n, x, norm = True, diff_n = 1)` will use `norm` and `diff_n` to select a ufunc that computes the normalised Legendre polynomials and their derivative. Additional positional arguments that precede the array arguments can be used to allocate the output, as `lpmn_all(m, n, x, norm = True, diff_n = 1)` needs to allocate output arrays of shape `(2 m + 1, n, ...)`.
- There are no "special" keywords, it is just convention. So `diff_n` has no special meaning and is not builtin to the `multiufunc` class or anything like that.
- You can see how this works for `lpmn` at https://github.com/scipy/scipy/blob/79421615dc66425bc47899f2afe254f5c7284398/scipy/special/_basic.py#L1727

I know this is a lot of information. I've been thinking and working on how to solve these problems for a long time. My guiding principle has been to _keep the fundamental model of computation a ufunc_ and build on top of that. Happy to discuss and elaborate more.

[1] https://github.com/scipy/scipy/issues/19404
[2] https://github.com/scipy/scipy/pull/20319
[3] https://github.com/jishnub/LegendrePolynomials.jl; https://www.mathworks.com/help/matlab/ref/legendre.html
[4] https://github.com/scipy/scipy/issues/20448; https://discuss.scientific-python.org/t/naming-convention-for-generalized-ufuncs-in-special/1144;

izaid commented:
This PR is, in fact, ready for wider discussion. I'm going to update my original post to explain it.

Commit:
.

Commit:
.

Commit:
.

Commit:
changing number of recurrence terms

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

izaid commented:
Right, so this is done now. I've done a lot more refactoring at the C++ level to get things where I want them to be, but haven't changed anything at the Python level since I last posted. I still haven't implemented normalized Legendre polynomials, but the interface is there for them, and they will come in a follow-up PR.

izaid commented:
I also want to highlight again the `diff_n` keyword we are introducing. This is being used to select how many derivatives (0, 1, or 2) that should be returned. The intention is to make things more consistent in `special`. Currently, throughout `special`, there are many functions that have to return the value and the first derivative, without the ability to not return the derivative. There are also a couple of other ad-hoc arrangements (like `spherical_jn`, which has a `derivative` keyword). By introducing `diff_n` we are working towards, if not a standard, at least a convention for how to handle returning derivatives or not. And one that supports higher orders than first (indeed, here I've implemented up to second order here).

The alternative to a `diff_n` is to make functions like `lpmn_derivative` or `lpmn_derivative2`, which doesn't make sense to me. You could make these functions methods on the custom ufunc wrapper, but it's still not great. Even though `diff_n` changes the number of outputs, I think it is nicest thing that can be done here.

Again, happy to hear different names for `diff_n` if anyone has them.

cc @steppi

lucascolley commented on scipy/special/_multiufunc.py:
```diff
@@ -0,0 +1,61 @@
+import numpy as np
+
+class multiufunc:
```
somebody else mentioned it elsewhere, but [PEP8 says](https://peps.python.org/pep-0008/#class-names) to style this as `MultiUfunc` (or `Multiufunc`, if it's supposed to be one word)

izaid replied:
Indeed, thanks @lucascolley! I have renamed `multiufunc` to `MultiUFunc`.

mdhaber commented:
Sorry for the driveby, but just throwing something out there in case it wasn't considered: an array of the derivative orders that is just broadcast with the other arguments.

Commit:
renamed multiufunc to MultiUFunc

izaid commented:
> Sorry for the driveby, but just throwing something out there in case it wasn't considered: instead of an integer `diff_n`, pass an array of the derivative orders that is just broadcast with the other arguments. So it would be e.g. `[0, 2]` to get the function value and second derivative. I'm not sure if there is any computational savings, but then the number of outputs doesn't depend on an integer argument; it would just follow broadcasting rules.

Thanks for the thoughts @mdhaber! That is a good suggestion. I'm all for _allowing_ `diff_n` to be able to do that, and indeed there is nothing here forcing `diff_n` to be an integer or anything like that. So totally agree there.

For the Legendre polynomials in particular, it's just not possible to do that unfortunately. Essentially the first derivative of the Legendre polynomials requires computing the value, the second derivative requires computing the first derivative, and so on.

mdhaber commented:
> it's just not possible to do that unfortunately.

Sure - i mentioned that there might not be computational savings. But could you only *return* the orders that are requested?

izaid commented:
> > it's just not possible to do that unfortunately.
> 
> Sure - i mentioned that there might not be computational savings. But could you only _return_ the orders that are requested?

Gotcha, understood.

All `diff_n` is is a keyword that selects _which_ ufunc to use. In this case, it would be the ufunc that returns only the value (`diff_n = 0`), the ufunc that returns the value and derivative (`diff_n` = 1), or the ufunc that returns the 1st derivative, 2nd derivative, and the value (`diff_n` = 2). The keyword `diff_n` is not builtin or anything, it's just a convention I'd like to see `special` stick to. For functions that want custom behaviour, like you suggested, perhaps they should just apply decorators to select appropriate outputs or whatever.

Commit:
Merge remote-tracking branch 'origin/main' into legendre

Commit:
enabled lambdas for special functions

Commit:
more lambda stuff

Commit:
templated clpmn_all

Commit:
Covered all norm and unnorm cases

Commit:
recurrence updates

Commit:
merged from main

Commit:
.

Commit:
.

Commit:
Py_BuildValue updates

Commit:
added normalised legendre polynomials

Commit:
added norm to clpmn tests

Commit:
cleaning up tests

Commit:
renaming of legendre functions

Commit:
renamed some type to typ

Commit:
.

Commit:
more clean-up

Commit:
.

Commit:
.

Commit:
.

Commit:
refactored out specific assoc_legendre tests

Commit:
fix for litner

Commit:
more lint fixes

Commit:
switched assoc_legendre_p n and m arg order

Commit:
.

Commit:
renaming clpmn to multi_assoc_legendre_p

Commit:
.

Commit:
permuted n and m dimensions in legendre

Commit:
fixes for lint

Commit:
.

Commit:
.

Commit:
renamed legendre test functions

Commit:
reorganising some tests

Commit:
.

Commit:
.

Commit:
.

Commit:
testing experimental grad_tuple

Commit:
more grad_tuple testing

Commit:
moved all of legendre_p over

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
grad refactoring

Commit:
refactoring assoc_legendre for grad

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
refactoring grad<...>

Commit:
refactoring grad still

Commit:
refactoring for refs

Commit:
more refactoring

Commit:
removed refs as tuple functions

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
.

Commit:
turning grad into a std::tuple where possible

Commit:
renamed grad to tuple_wrapper

Commit:
renamed grad.h to tuple_wrapper.h

Commit:
.

Commit:
converted refs to apply_tie

Commit:
worked out tuple_assigner

Commit:
cleaned up tuple_assigner

Commit:
.

Commit:
.

Commit:
bit more refactoring

Commit:
.

Commit:
added grad_tuple_t

Commit:
.

Commit:
removed diag bool flag

Commit:
.

Commit:
.

Commit:
various refactoring and renaming

Commit:
removed unnecessary includes from sph_harm

Commit:
added sph_legendre and updated sph_harm

Commit:
Turned tuple helper functions into their own namespace

Commit:
.

Commit:
slight dot refactoring

Commit:
added sph_harm_y_next

Commit:
Fully added sph_legendre_p

Commit:
fix for linter

Commit:
working towards sph_harm derivatives

Commit:
added derivatives for sph_harm_y

steppi committed:
Make multiufunc attributes public + adjust names

steppi committed:
Use slots in MultiUFunc

Commit:
made sph_harm_y_all a multiufunc

steppi committed:
Add signature for resolve_ufunc and make not takes ufuncs

Commit:
fix for linter

steppi committed:
Adjust name/docs of resolution methods

Commit:
.

steppi committed:
Add input validation to multiufunc.

steppi committed:
Use distinct names for ufuncs within multiufunc

steppi committed:
Begin documenting MultiUFunc

steppi committed:
Only allow resolvers set by decorator

- There should be one-- and preferably only one --obvious way to do it.

steppi committed:
preliminary work to set docstrings

steppi committed:
Get instance level docstring working

steppi committed:
Make separate docstring and class docstring work

steppi committed:
Remove hack in of class docstring for MultiUFunc

steppi committed:
Add new multifuncs to docs

Commit:
rebased

Commit:
Merge remote-tracking branch 'origin/main' into legendre

steppi committed:
Clean up all merge conflicts

- for real this time

steppi committed:
Add input validation for diff_n

steppi committed:
Update docs for MultiUFuncs

steppi committed:
Docstring for assoc_legendre_p

steppi committed:
Get rid of the __slots__ in MultiUFunc

steppi committed:
Separate _multiufuncs out of _basic

steppi committed:
Use _nonneg_int_or_fail to validate diff_n

steppi committed:
Update Legendre docs

steppi committed:
Update docstring short descriptions for consistency

steppi committed:
Pointless thrashing

izaid committed:
Merge pull request #2 from steppi/multiufunc-docs

Thanks @steppi! Merging after discussion with you.

Commit:
Made docs a settable attribute from the constructor

Commit:
removed unused import of operator

Commit:
made names the same again

Commit:
added missing r

Commit:
added _multiufuncs to mypy

Commit:
forces_complex_output -> force_complex_output

Commit:
switched as_ prefix to register_

Commit:
reworked some docs, working on sph_harm hess

Commit:
fix for linter

Commit:
Some doc updates

Commit:
added map_dims to sort out sph_harm_all derivatives

Commit:
deleted steps_offset, made ranks_scans

Commit:
synced non-gufuncs

Commit:
.

Commit:
trying axes keyword of gufunc

Commit:
migrated multiufunc fully to use axes keyword

Commit:
cleaned up all test

Commit:
made resolve_ufunc return a key

Commit:
flattened collections of ufuncs

Commit:
.

Commit:
made typ a default argument for multiufunc

Commit:
removed assoc_legendre_p_all

Commit:
deleted assoc_legendre_p

Commit:
removed force_complex for multi_assoc_...

Commit:
.

Commit:
Revert "deleted assoc_legendre_p"

This reverts commit 560b075b014d4a48733c7811340d3291d0ce413e.

Commit:
Revert "removed assoc_legendre_p_all"

This reverts commit 2063e469dbc939ffd0ad71bc4220bba368e9d303.

Commit:
Revert "made typ a default argument for multiufunc"

This reverts commit a50ce036d78ceaf4cb0c63b1442d057ebc56d94a.

Commit:
readded force_complex_output flag

Commit:
switched order of z and type for multi_assoc_legendre_p and multi_assoc_legendre_p_all

Commit:
made lpmn use multi_assoc_legendre_p_all

Commit:
added real overloads

Commit:
removed force_complex_output from multi_assoc_legendre_p_all

Commit:
added float overloads for multi_assoc_legendre_p

Commit:
added cast to complex in clpmn

Commit:
removed force_complex_output from multi_assoc_legendre_p

Commit:
moved several tests to multi_*

Commit:
removed assoc_legendre_p from Python

Commit:
removed assoc_legendre_p_* and assoc_legendre_p_all_*

Commit:
added typ keyword to multi_legendre_p_all

Commit:
added typ keyword to multi_legendre_p

Commit:
added default values to typ

Commit:
removed assoc_legendre C++ code

Commit:
made MultiUFunc method setters consistently called override, renamed multi_ to assoc_

Commit:
.

Commit:
.

Commit:
renamed all C++ instances of multi_

Commit:
synced testing names

Commit:
switched order of type and z in global namespace

Commit:
swapped more z and type in C++

Commit:
.

Commit:
added a default kwargs function to MultiUFunc

Commit:
multiufunc refactoring

Commit:
fix for linter

Commit:
.

Commit:
.

Commit:
Merge remote-tracking branch 'origin/main' into legendre

steppi commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+        return self.__doc
+
+    @property
+    def force_complex_output(self):
+        return self.__force_complex_output
```
We discussed this offline, but I'm writing it here to make sure it doesn't get forgotten in the shuffle. `force_complex_output` should be private since the plan is to remove it when 1.24 becomes the minimum supported NumPy version and we can use `resolve_dtypes`.

steppi commented on scipy/special/_basic.py:
```diff
@@ -3415,22 +3358,4 @@ def zeta(x, q=None, out=None):
-    _sph_harm_all_gufunc(theta, phi, out = np.moveaxis(out, (0, 1), (-2, -1)))
-
-    return out
+        return _ufuncs._zeta(x, q, out)
```
I think we should configure the lint CI job to check for blank lines at the end of files.

lucascolley replied:
done in gh-21023

steppi commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+        func.__name__ = "resolve_out_shapes"
+        self.resolve_out_shapes = func
+
+    def resolve_ufunc(self, **kwargs):
```
Before your latest round of commits, I had made it so that `resolve_ufunc` had an accurate signature. Is this just an inadvertent change? I'd prefer to keep the accurate signatures if possible. Right now on your branch we have:

```
In [10]: assoc_legendre_p_all.resolve_ufunc?
Signature: assoc_legendre_p_all.resolve_ufunc(**kwargs)
Docstring: Resolve to a ufunc based on keyword arguments.
File:      ~/scipy/build-install/lib/python3.12/site-packages/scipy/special/_multiufuncs.py
Type:      method
```

but on an earlier commit this was

```
In [2]: assoc_legendre_p_all.resolve_ufunc?
Signature: assoc_legendre_p_all.resolve_ufunc(norm=False, diff_n=0)
Docstring: Resolve to ufunc dispatched to for given keyword arguments.
File:      ~/scipy/build-install/lib/python3.12/site-packages/scipy/special/_multiufuncs.py
Type:      function
```

izaid replied:
We can do this is required. But what changed is that `resolve_ufunc` is no longer something the user implements, it just calls the `key` function. It itself never changes.

To do what you are suggesting would require monkeypatching it, but we have no reason to do that anymore. I say we leave it as is for simplicity. Again, though, happy to do what you are suggesting for any other relevant functions.

steppi replied:
> What changed is that `resolve_ufunc` is no longer something the user implements, it just calls the `key` function. It itself never changes.

I think there's a distinction we need to make clear. The users of `MultiUFunc` are the people who import the special functions and use them in their code. Users were never required to implement `resolve_ufunc`, what we're discussing is the internal API for setting up `MultiUFunc`s for the user. Monkeypatching is no longer required to get the functionality to work correctly, but having a proper signature is very helpful for the user, especially because some of a `MultiUFunc`s kwargs are used to give ufuncs default args and some are used to choose which ufunc to dispatch to. I think our main consideration in designing `MultiUFunc` should be the user facing API, not how we set things up internally.

izaid replied:
Ok, we've come to a compromise that `resolve_ufunc` shall be private for now. I have turned it into `_resolve_ufunc`.

steppi commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+
+    Parameters
+    ----------
+    n : array_like, int
+        Order (max) of the Legendre polynomial, must have ``n >= 0``.
+    m : array_like, int
+        Degree (max) of the Legendre polynomial, must have ``m >= 0``
```
I'd mentioned this offline, but for the `all` functions, when true, we need to document that the tables are for all orders $0 \leq i \leq n$ and all degrees $-m \leq j \leq m$, and that the values are packed into the arrays in such a way that negative indexing works as expected.

izaid replied:
Have done so.

steppi reviewed (commented):
Hey Irwin, things are looking good. I have some comments, mostly items that we discussed offline that I've put here to make sure we don't forget about them.

izaid commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+        return self.__doc
+
+    @property
+    def force_complex_output(self):
+        return self.__force_complex_output
```
Happy with this.

izaid commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+
+        ufunc = self.resolve_ufunc(**kwargs)
+
+        ufunc_args = args[-ufunc.nin:] # array arguments to be passed to the ufunc
```
Happy with this. It's interesting _why_ array arguments don't seem to work for `_all` whereas they do for the others. I think using the `axes` keyword of ufuncs somehow prevents the internal `asarray` from being called. Something to talk to NumPy about!

lucascolley commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+        Degree of the Legendre polynomial.
+    z : array_like, float
+        Input value.
+    typ : Optional[int]
```
+1 for a more descriptive name

izaid replied:
Ok, some history here. Previously there was an argument called `type` here. What I did so far was rename it to `typ` to prevent clashing. I actually have never used this argument. My online searching indicates nearly all the associated Legendre polynomial implementations I can find don't have this argument.

Of course, it's valuable, so we shouldn't remove it or anything. Mathematica _does_ have this argument and calls it "type". Maple also has an option for this and call is "LegendreCut" [2].

Is there some reason not to just call it `branch_cut` or `cut` like Maple?

[1] https://mathoverflow.net/questions/62358/legendre-function-types
[2] https://www.maplesoft.com/support/help/maple/view.aspx?path=Legendre

steppi replied:
I could accept just `branch_cut`. I suggested `branch_cut_type` for more continuity with the previous name.

izaid replied:
Have renamed it to `branch_cut`.

lucascolley commented on scipy/special/tests/test_legendre.py:
```diff
@@ -0,0 +1,1476 @@
+def sph_legendre_p_3_1(phi):
+    fac = sph_legendre_factor(3, 1)
+
+    return -3 * fac * (5 * np.square(np.cos(phi)) - 1) * \
+        np.abs(np.sin(phi)) / 2
```
Totally optional style opinion: not a fan of the `\`'s everywhere.

```suggestion
    res = (-3 * fac * (5 * np.square(np.cos(phi)) - 1)
           * np.abs(np.sin(phi)) / 2)
    return res
```

izaid replied:
Out of curiousity, what's your preferred alternative? Just brackets `(...)`?

steppi replied:
Yeah, just brackets. From Pep8 again

```
The preferred way of wrapping long lines is by using Python’s implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.
```

izaid replied:
Fine, happy with that.

lucascolley reviewed (commented):
some drive by comments!

lucascolley commented on scipy/special/tests/test_legendre.py:
```diff
@@ -0,0 +1,1476 @@
+        m = rng.integers(-10, 10, shape)
+        x = rng.uniform(-1, 1, shape)
+
+        p, p_jac, p_hess = special.assoc_legendre_p(n, m, x, norm = norm, diff_n = 2)
```
There is a ruff rule, https://docs.astral.sh/ruff/rules/unexpected-spaces-around-keyword-parameter-equals/, but it is unstable, so we won't add it to the linter yet. You can use ruff to fix up this PR though, with `ruff check --select E251 --preview --fix .`.

izaid replied:
Do you guys have a hard rule on this? I personally would normally use keywords like `some_keyword = ...`. This works well especially when the keyword has a more complicated value, such as `some_keyword = [2 * i for i in range(10)]`.

But can make those changes if needed.

steppi replied:
Yes, this is a hard rule. Having whitespace in these situations is very unorthodox for Python.

izaid replied:
So, for instance, should I be writing `axis=(0,)+tuple(range(2,theta.ndim+2))` or `axis = (0,) + tuple(range(2, theta.ndim + 2))`? If you really want the former, I can, looks bad to me.

lucascolley replied:
I feel like that example should almost certainly be

```python
axis = (0,) + tuple(range(2, theta.ndim + 2))
func(..., axis=axis, ...)
```

izaid replied:
So, on reflection, I think I'm one of those people that doesn't fully agree with PEP8, for some of the reasons discussed here: https://stackoverflow.com/questions/8853063/pep-8-why-no-spaces-around-in-keyword-argument-or-a-default-parameter-value

But we can do it like you suggest, I'm happy if SciPy is happy.

izaid commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+    def force_complex_output(self):
+        return self.__force_complex_output
+
+    def override_key(self, func):
+        """Set `key` method by decorating a function.
+        """
+        self.key = func
+
+    def override_ufunc_default_args(self, func):
+        self.ufunc_default_args = func
+
+    def override_ufunc_default_kwargs(self, func):
+        self.ufunc_default_kwargs = func
+
+    def override_resolve_out_shapes(self, func):
+        """Set `resolve_out_shapes` method by decorating a function."""
+        if func.__doc__ is None:
+            func.__doc__ = \
+                """Resolve to output shapes based on relevant inputs."""
+        func.__name__ = "resolve_out_shapes"
+        self.resolve_out_shapes = func
```
This I disagree on. We want MultiUFunc to look nice, including to us. Prefixing the core methods one needs to use with an underscore doesn't achieve that. Also, a method is clearly not private if it needs to be used outside the class.

I'd like to keep this as is. The class MultiUFunc is in a private module, so I think it already is private from a SciPy perspective.

steppi replied:
Ok. But each individual MultiUFunc will be public and will have these methods, but they are only used for setting things up, they will not or at least should not be used by the end users of the MultiUFuncs. That they exist is only an implementation detail that does not need to be “known” by users of `assoc_legendre_p`

izaid replied:
Agree! Our end users should treat these objects like a ufunc, as much as we can support that.

izaid commented on scipy/special/_basic.py:
```diff
@@ -1762,38 +1749,31 @@ def lpmn(m, n, z):
 
 
-def clpmn(m, n, z, type=3):
+def clpmn(m, n, z, type = 3):
```
Ok.

izaid commented on scipy/special/_gufuncs_docs.cpp:
```diff
@@ -231,8 +230,4 @@ const char *rcty_doc = R"(
-const char *sph_harm_all_doc = R"(
-    Private function. This may be removed or modified at any time.
-    )";
+    )";
```
This is a good point that I have caught too. Essentially the issue is we have a `.clang-format` in the `special/special` subdirectory, as that is what will soon get spun out. In the parent directory, this doesn't apply. Probably SciPy itself needs a global `.clang-format` format somewhere, which may just be a copy of the one in `special/special`.

izaid commented on scipy/special/_input_validation.py:
```diff
@@ -0,0 +1,19 @@
+import operator
+
+from numpy import floor
+
+
+def _nonneg_int_or_fail(n, var_name, strict=True):
+    try:
+        if strict:
+            # Raises an exception if float
+            n = operator.index(n)
+        elif n == floor(n):
```
Yeah, actually I don't know we aren't just using math.floor here. We just inherited `_nonneg_int_or_fail` from somewhere, but it will never be an array argument. Will use math.floor.

steppi replied:
Yeah, this was written in the early days and was in `_basic.py`. We just separated it out to avoid a circular dependency.

izaid replied:
Have done this.

izaid commented on scipy/special/__init__.py:
```diff
@@ -470,17 +470,30 @@
 .. autosummary::
    :toctree: generated/
 
-   lpmv     -- Associated Legendre function of integer order and real degree.
```
Have done this.

steppi commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+    Compute sequence of Legendre functions of the first kind (polynomials),
+    Pn(z) and derivatives for all degrees from 0 to n (inclusive).
+
+    See also special.legendre for polynomial class.
```
Just pointing out this was copied verbatim from the original `lpn` docstring. This is a good opportunity to fix it.

izaid replied:
Done!

izaid commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+
+    This is the same as the spherical harmonic :math:`Y_{m}^{n}(\theta, \phi)`
+    with :math:`\theta = 0`.
+    """, diff_n=0
```
I prefer it like this tbh, but can do so if SciPy has a rule.

steppi replied:
My preference is the same as @lucascolley's, but I don't think there's a rule on this.

izaid replied:
No rule then I leave it as is. 😈

izaid commented on scipy/special/_multiufuncs.py:
```diff
@@ -0,0 +1,640 @@
+                ufunc_out_dtypes = tuple(np.result_type(1j, ufunc_out_dtype)
+                    for ufunc_out_dtype in ufunc_out_dtypes)
+
+            out = tuple(np.empty(ufunc_out_shape, dtype = ufunc_out_dtype)
```
Done!

Commit:
several fixes and docs

izaid commented on scipy/special/tests/test_legendre.py:
```diff
@@ -0,0 +1,1476 @@
+
+import pytest
+from numpy.testing import (assert_equal, assert_almost_equal, assert_array_almost_equal,
+    assert_, assert_allclose)
```
Sure, those tests predated me but changed anyway.

izaid commented on scipy/special/_basic.py:
```diff
@@ -2041,38 +2013,9 @@ def euler(n):
-    .. [1] Zhang, Shanjie and Jin, Jianming. "Computation of Special
-           Functions", John Wiley and Sons, 1996.
-           https://people.sc.fsu.edu/~jburkardt/f77_src/special_functions/special_functions.html
-
```
Put back docstring for `lpn` and added deprecated notices!

Commit:
added deprecation notices to docs

Commit:
updated sph_harm_y and sph_harm_y_all docs

Commit:
fixes for linter

izaid commented on scipy/special/tests/test_legendre.py:
```diff
@@ -0,0 +1,1476 @@
+from numpy.testing import (assert_equal, assert_almost_equal, assert_array_almost_equal,
+    assert_, assert_allclose)
+
+from scipy import special
```
I have done so for all tests using the new functions. Happy to adapt tests using the old functions as we go.

Commit:
.

izaid commented:
@steppi @lucascolley I have now addressed all your comments. In a couple of places I have left questions for you, notably on the `typ` / `branch_cut` argument name.

I'd very much appreciate it if you could have another look through this as soon as possible. It's a complex and big PR, so good to as many eyes on it as possible.

Commit:
clean-up

Commit:
made resolve_ufunc private

Commit:
renamed typ to branch_cut

izaid commented:
I have renamed `typ` to `branch_cut`.

Commit:
simplified the "all" functions documentation

Commit:
fix for linter

Commit:
renamed type -> branch_cut in some of the C++

Commit:
renamed rest of C++ code from type -> branch_cut

Commit:
various Python refactoring

izaid commented:
I think I've addressed everything, cc @steppi

Commit:
Revert "various Python refactoring"

This reverts commit c50cfd194cadbeaec4ab875a8c68bdcc020c22f3.

Commit:
reverted

Commit:
updated default testing arguments

Commit:
removed unnecessary keyword use

Commit:
renamed typ_sign to branch_cut_sign

Commit:
updated exception

Commit:
.

Commit:
.

steppi commented:
The test failure

```
ERROR: objects in refguide but not in scipy.special.__all__::

    assoc_legendre_p
    assoc_legendre_p_all
    legendre_p
    legendre_p_all
    sph_harm_y
    sph_harm_y_all
    sph_legendre_p
    sph_legendre_p_all

This issue should likely be fixed by removing these objects
from the function listing in __init__.py for this module
or adding them to __all__.
```

can be fixed by changing this line:

https://github.com/scipy/scipy/blob/72c5707b1796164ca2e6cb85f0399d66d61cca3d/scipy/special/__init__.py#L857

to:

```python
__all__ = _ufuncs.__all__ + _basic.__all__ + _orthogonal.__all__ + _multiufuncs.__all__
```

steppi commented:
You can get rid of this warning on gcc

```
../scipy/special/special/recur.h:89:6: note: the ABI of passing structure with ‘complex float’ member has changed in GCC 4.4
   89 | void forward_recur(InputIt first, InputIt last, Recurrence r, std::tuple<T (&)[K]...> res, Func f) {
```

by adding `-Wno-psabi` to the `cpp_args` for the `_gufuncs` and `_special_ufuncs` extension modules in special's meson.build ([from stackoverflow](https://stackoverflow.com/questions/13913791/remove-note-of-gcc-abi-change)). Since these extension modules will never need to interact with programs compiled with an earlier GCC, this ABI changes won't affect us.

steppi commented:
@izaid, I'm going to go ahead and make the small changes mentioned above myself and push since I know you're busy now.

steppi committed:
Suppress ABI change warning from special/recur.h

- ../scipy/special/special/recur.h:89:6: note: the ABI of passing structure with ‘complex float’ member has changed in GCC 4.4

steppi committed:
Add multiufuncs to special's __all__

steppi commented:
The multiufuncs still have a lot of scaffolding methods that are not meant to be user facing which are still public. I'll push an update for this too.

```
 'key',
 'override_key',
 'override_resolve_out_shapes',
 'override_ufunc_default_args',
 'override_ufunc_default_kwargs',
 'resolve_out_shapes',
 'ufunc_default_args',
 'ufunc_default_kwargs'
```

steppi committed:
Make all multiufunc scaffolding private

Commit:
switched meaning of theta and phi in new spherical harmonics

Commit:
updated new sph_harm docs and made angle change consistent for sph_legendre_p

Commit:
removed underscores from override methods

Commit:
Merge remote-tracking branch 'origin/main' into legendre

Commit:
removed underscores from override methods

steppi commented on scipy/special/_basic.py:
```diff
@@ -2041,7 +2027,7 @@ def euler(n):
     return _specfun.eulerb(n1)[:(n+1)]
 
-
+    
```
```suggestion

```

steppi commented on scipy/special/_gufuncs_docs.cpp:
```diff
@@ -231,8 +230,4 @@ const char *rcty_doc = R"(
-const char *sph_harm_all_doc = R"(
-    Private function. This may be removed or modified at any time.
-    )";
+    )";
```
```suggestion
    )";

```

steppi reviewed (commented):
The windows with MSVC failure is real,

```
2024-07-08T09:39:50.6342960Z FAILED: scipy/special/_special_ufuncs.cp311-win_amd64.pyd.p/_special_ufuncs.cpp.obj 
2024-07-08T09:39:51.4336133Z "cl" "-Iscipy\special\_special_ufuncs.cp311-win_amd64.pyd.p" "-Iscipy\special" "-I..\scipy\special" "-Iscipy\_lib" "-I..\scipy\_lib" "-I..\scipy\_build_utils\src" "-IC:\Miniconda\envs\__setup_conda\Lib\site-packages\numpy\_core\include" "-IC:\Miniconda\envs\__setup_conda\Include" "/MD" "/nologo" "/showIncludes" "/utf-8" "/Zc:__cplusplus" "/W2" "/EHsc" "/std:c++17" "/permissive-" "/O2" "/Zi" "-DMS_WIN64=" "-DSP_SPECFUN_ERROR" "-DSCIPY_DLL_IMPORTS" "-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION" "-Wno-psabi" "/Fdscipy\special\_special_ufuncs.cp311-win_amd64.pyd.p\_special_ufuncs.cpp.pdb" /Foscipy/special/_special_ufuncs.cp311-win_amd64.pyd.p/_special_ufuncs.cpp.obj "/c" ../scipy/special/_special_ufuncs.cpp
2024-07-08T09:39:51.4339417Z cl : Command line error D8021 : invalid numeric argument '/Wno-psabi'
```

The `Wno-psabi` argument needs to be added conditionally only for GCC, I'll take care of it and the formatting changes. I have a windows machine at home to test this on.

steppi committed:
Only add -Wno-psabi when compiler is gcc

steppi committed:
Add missing newline

steppi committed:
Remove extraneous whitespace

steppi committed:
Formatting changes for pep8

steppi commented:
We've got more errors in the Windows + MSVC job that were masked by the previous error.

```
2024-07-08T18:42:19.8664491Z D:\a\scipy\scipy\scipy\special\special\recur.h(88): error C2061: syntax error: identifier 'ssize_t'
2024-07-08T18:42:19.8665354Z D:\a\scipy\scipy\scipy\special\special\recur.h(89): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8666200Z D:\a\scipy\scipy\scipy\special\special\recur.h(89): error C2061: syntax error: identifier 'Func'
2024-07-08T18:42:19.8667045Z D:\a\scipy\scipy\scipy\special\special\recur.h(91): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8667913Z D:\a\scipy\scipy\scipy\special\special\recur.h(98): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8668723Z D:\a\scipy\scipy\scipy\special\special\recur.h(100): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8669537Z D:\a\scipy\scipy\scipy\special\special\recur.h(107): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8670353Z D:\a\scipy\scipy\scipy\special\special\recur.h(107): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8671157Z D:\a\scipy\scipy\scipy\special\special\recur.h(91): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8671954Z D:\a\scipy\scipy\scipy\special\special\recur.h(98): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8672747Z D:\a\scipy\scipy\scipy\special\special\recur.h(100): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8673705Z D:\a\scipy\scipy\scipy\special\special\recur.h(115): error C2061: syntax error: identifier 'ssize_t'
2024-07-08T18:42:19.8674556Z D:\a\scipy\scipy\scipy\special\special\recur.h(116): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8675400Z D:\a\scipy\scipy\scipy\special\special\recur.h(116): error C2061: syntax error: identifier 'Func'
2024-07-08T18:42:19.8676239Z D:\a\scipy\scipy\scipy\special\special\recur.h(118): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8677046Z D:\a\scipy\scipy\scipy\special\special\recur.h(125): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8677840Z D:\a\scipy\scipy\scipy\special\special\recur.h(127): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8678641Z D:\a\scipy\scipy\scipy\special\special\recur.h(134): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8679445Z D:\a\scipy\scipy\scipy\special\special\recur.h(134): error C3861: 'K': identifier not found
2024-07-08T18:42:19.8680246Z D:\a\scipy\scipy\scipy\special\special\recur.h(118): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8681056Z D:\a\scipy\scipy\scipy\special\special\recur.h(125): error C2065: 'K': undeclared identifier
2024-07-08T18:42:19.8682327Z D:\a\scipy\scipy\scipy\special\special\recur.h(127): error C2065: 'K': undeclared identifier
```

The issue is that `ssize_t` is POSIX but not C++ standard. It seems the usual solution is to use `ptrdiff_t` as a substitute as described here, https://www.scivision.dev/ssize_t-platform-independent/. I can take care of this too.

That Windows + MSVC CI job is a real lifesaver.

steppi committed:
More pep8 formatting updates

steppi committed:
Replace ssize_t (not standard) with ptrdiff_t

steppi commented:
More Windows + MSVC failures. My local Windows + MSVC build is erroring out for unrelated reasons before it gets to `special`. I'll need to fix that so I can debug this properly.

izaid commented:
> More Windows + MSVC failures. My local Windows + MSVC build is erroring out for unrelated reasons before it gets to `special`. I'll need to fix that so I can debug this properly.

Ok, thanks for looking into this! I'm a bit surprised, as all the Windows CI passes, but there you go. Appreciate it.

steppi commented:
It turns out I can just use gcc with the `--pedantic` flag and check the warnings to debug locally. The primary thing I'm seeing leading to failures is that MSVC doesn't allow for creation of zero-size arrays, which isn't supported by the C++ standard.

izaid commented:
> It turns out I can just use gcc with the `--pedantic` flag and check the warnings to debug locally. The primary thing I'm seeing leading to failures is that MSVC doesn't allow for creation of zero-size arrays, which isn't supported by the C++ standard.

Where is that? Probably an easy fix, just ping if you want to discuss.

steppi commented:
@izaid, I won't be available for the next few hours, but you can search for error in the logs here or add pedantic to your `CXXFLAGS` before compiling locally and check the warnings

```
export CXXFLAGS="--pedantic"                                                                                                                
python dev.py build
```

I think it is an easy fix. If you have time to fix it, that would be great, otherwise I can do it tonight.

Commit:
fix for zero-size arrays

izaid commented:
@steppi Done.

steppi commented:
Another one

```
2024-07-08T21:56:45.2319737Z D:\a\scipy\scipy\scipy\special\special/legendre.h(332): error C2668: 'signbit': ambiguous call to overloaded function
2024-07-08T21:56:45.2320871Z C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(319): note: could be 'bool signbit(long double) noexcept'
2024-07-08T21:56:45.2322051Z C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(314): note: or       'bool signbit(double) noexcept'
2024-07-08T21:56:45.2323191Z C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(309): note: or       'bool signbit(float) noexcept'
2024-07-08T21:56:45.2324318Z D:\a\scipy\scipy\scipy\special\special/legendre.h(332): note: while trying to match the argument list '(int)'
2024-07-08T21:56:45.2325284Z D:\a\scipy\scipy\scipy\special\special/legendre.h(332): note: the template instantiation context (the oldest one first) is
2024-07-08T21:56:45.2326943Z ../scipy/special/_special_ufuncs.cpp(546): note: see reference to function template instantiation 'void `anonymous-namespace'::assoc_legendre_p<special::assoc_legendre_norm_policy,double>(NormPolicy,__int64,__int64,T,__int64,T &,T &)' being compiled
```

and `--pedantic` doesn't catch this one on gcc.

The offending lines are

```C++
template <typename NormPolicy, typename T, typename... OutputVals, typename Func>
void assoc_legendre_p_for_each_m_abs_m(
    NormPolicy norm, int m, T z, int branch_cut, std::tuple<OutputVals (&)[2]...> res, Func f
) {
    bool m_signbit = std::signbit(m);
```

which is kind of weird. The [signbit docs](https://en.cppreference.com/w/cpp/numeric/math/signbit)  say that `std::signbit` has an overload for `int`s as of C++11, but MSVC doesn't support this https://github.com/microsoft/STL/issues/519. Although it looks like that's going to change in a coming version, https://github.com/microsoft/STL/pull/4537.

I guess just use `m < 0`

Commit:
removed signbit call

izaid commented:
> Another one
> 
> ```
> 2024-07-08T21:56:45.2319737Z D:\a\scipy\scipy\scipy\special\special/legendre.h(332): error C2668: 'signbit': ambiguous call to overloaded function
> 2024-07-08T21:56:45.2320871Z C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(319): note: could be 'bool signbit(long double) noexcept'
> 2024-07-08T21:56:45.2322051Z C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(314): note: or       'bool signbit(double) noexcept'
> 2024-07-08T21:56:45.2323191Z C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt\corecrt_math.h(309): note: or       'bool signbit(float) noexcept'
> 2024-07-08T21:56:45.2324318Z D:\a\scipy\scipy\scipy\special\special/legendre.h(332): note: while trying to match the argument list '(int)'
> 2024-07-08T21:56:45.2325284Z D:\a\scipy\scipy\scipy\special\special/legendre.h(332): note: the template instantiation context (the oldest one first) is
> 2024-07-08T21:56:45.2326943Z ../scipy/special/_special_ufuncs.cpp(546): note: see reference to function template instantiation 'void `anonymous-namespace'::assoc_legendre_p<special::assoc_legendre_norm_policy,double>(NormPolicy,__int64,__int64,T,__int64,T &,T &)' being compiled
> ```
> 
> and `--pedantic` doesn't catch this one on gcc.
> 
> The offending lines are
> 
> ```c++
> template <typename NormPolicy, typename T, typename... OutputVals, typename Func>
> void assoc_legendre_p_for_each_m_abs_m(
>     NormPolicy norm, int m, T z, int branch_cut, std::tuple<OutputVals (&)[2]...> res, Func f
> ) {
>     bool m_signbit = std::signbit(m);
> ```
> 
> which is kind of weird. The [signbit docs](https://en.cppreference.com/w/cpp/numeric/math/signbit) say that `std::signbit` has an overload for `int`s as of C++11, but MSVC doesn't support this [microsoft/STL#519](https://github.com/microsoft/STL/issues/519). Although it looks like that's going to change in a coming version, [microsoft/STL#4537](https://github.com/microsoft/STL/pull/4537).
> 
> I guess just use `m < 0`

Fixed.

steppi reviewed (approved):
Phew. The jobs are finally all passing. Thanks @czgdp1807 for implementing the Windows + MSVC CI job. It's much better to catch these things in advance rather than find out about them just before a release.

I think this is really great work @izaid. Thank you for this and for your patience while I found bandwidth to review this properly. For everyone's sake I will summarize the changes, giving an emphasis to the enhancements and deprecations where we need to find consensus from maintainers. I will also describe some of the work that remains to be done here.

This PR proposes the following new functions, some of which replace existing functions which we propose to deprecate.

| New function               | Replaces    |
| --------------------------------| ----------------|
| legendre_p                   |  New          |
| legendre_p_all             |   lpn            |
| assoc_legendre_p       |     New        |
| assoc_legendre_p_all | lpmn, clpmn  |
| sph_harm_y                 | sph_harm  |
| sph_harm_y_all           |   New          |
| sph_legendre_p          |   New          |
| sph_legendre_p_all    |   New          |

New functions for Legendre functions of the second kind replacing `lqn` and `lqmn` will be added in a follow up PR.

Notice the existence of `_all` versions of each function. This was discussed on discourse here: https://discuss.scientific-python.org/t/naming-convention-for-generalized-ufuncs-in-special/1144, and on GitHub here: https://github.com/scipy/scipy/issues/20448.

To summarize, each of the functions in question take integral degree and order arguments `n` and `m`. The result of say `assoc_legendre_p(n, m, x)` is calculated based on a recurrence relation that computes `assoc_legendre_p(j, i, x)` for all $0 \leq j \leq n$, $0 \leq i \leq m$, $m \leq n$. Essentially, the `_all` version of each function takes only scalars for `m` and `n` and returns the entire table of values produced during the recurrence. The regular version is a ufunc which takes integer arrays for `m` and `n` and computes the recurrence separately each time the scalar kernel is called over array inputs. There are use cases for each. The `_all` versions avoid duplicate computation when one wants the table of values, while the regular versions will avoid excessive memory use if one only wants a handful of evaluations for large `n`. You may have noticed an inconsistency in the behavior of the existing functions. `lpmn`, `clpmn`, and `lpn` all have the `_all` behavior whereas `sph_harm` does not.

Something that should be pointed out is that for the original functions, the order `m` precedes the degree `n`, so the signature is `lpmn(m, n, z)`. The new functions now have the degree `n` before the order `m`, so the new function's signature is `assoc_legendre_p_all(n, m, z)`. This puts our functions in line with those in the [C++ stdlib](https://en.cppreference.com/w/cpp/numeric/special_functions/assoc_legendre), [Wolfram](https://reference.wolfram.com/language/ref/LegendreP.html), [Julia](https://juliapackages.com/p/legendrepolynomials) , [CERN ROOT](https://root.cern.ch/doc/master/classRooLegendre.html), and [mpmath](https://www.mathworks.com/help/matlab/ref/legendre.html). [MATLAB's behavior](https://www.mathworks.com/help/matlab/ref/legendre.html) is different in that it doesn't even have `m`, always returning the full table `0 <= i <= j <= n`. Since these are completely new functions, I think it's reasonable to switch the order of these arguments to be in line with those from other languages and packages.

There was a debate regarding whether it's better to have separate `_all` versions of each of these functions or to expose the `all` behavior behind a keyword argument, with separate functions winning out. Some considerations behind this design, the non `all` versions are ufuncs taking arrays for `m` and `n` with output shape depending on input shape, whereas the `_all` versions take only scalars for `m` and `n` and these parameters control the output shape . There are situations where the order and degree can take real or complex values (such as for the existing functiong `lpmv`), and the non-all ufuncs can and will be adapted to handle these cases in future work, whereas such a thing is not sensible for the `_all` functions. I think the general consensus is that having a keyword argument that is able to change the behavior of a function so dramatically is not desirable.

Note also that each of these new functions has a keyword argument, `diff_n`, which controls the number of derivatives to return. This is discussed in the PR message and @izaid informed maintainers of this on discourse here: https://discuss.scientific-python.org/t/naming-convention-for-generalized-ufuncs-in-special/1144/35?u=steppi.
To motivate this,  `lpn`, `lpmn`, `clpmn`, `lqn`, `lqmn` all return tuple of two arrays, the the element for the evaluation of a Legendre function and the second of its derivative. `sph_harm` only returns one array evaluating the function. For these functions, evaluating the $n^{th}$ derivative at a point requires evaluating the function and all derivatives of order up to $n-1$. When `diff_n` is set to `i`, the output of each of the proposed functions in this PR will be a tuple containing all derivatives up to order `i`. Currently only derivatives up to order 2 are supported. Returning a tuple of varying length based on a keyword argument is not ideal, but I've spoken to @izaid at length about this and now agree with him that it's an acceptable compromise. 

The reason for a tuple rather than a multidimensional array is in order to allow for multivariate functions such as `sph_harm_y`. If a function takes `k` input arguments, then the function itself evaluated at a scalar will yield a scalar, the gradient at a single point will be a vector of length `k`, the hessian an array of size $k\times k$, and so on for higher derivatives. These results could not be conveniently packed into a single array, so @izaid has gone with tuples like this. Having a keyword argument which controls the number of outputs isn't ideal, but it is valuable to be able to calculate these derivatives and to have control over how many derivatives are returned to avoid computing them when not needed.

Some other highlights. `assoc_legendre_p` and `assoc_legendre_p_all` have a `norm` keyword argument which controls whether or not to compute normalize the Legendre function by a scaling factor such that the $L^2$ norm over the domain is 1. Such an option is available in [Matlab](https://www.mathworks.com/help/matlab/ref/legendre.html), and is a useful addition in my opinion.

`assoc_legendre_p` and `assoc_legendre_p_all` each take both real and complex values. Before hand there were separate `lpmn` and `clpmn` for real and complex values respectively. The main issue causing the split is that there are multiple supported locations for the branch cut. `clpmn` had `type=2` for cuts along the real axis from $-1$ to $-\infty$ and $1$ to $\infty$ and `type=3` for a single cut along the real axis from $-1$ to $1$. `lpmn` extended the function to the entire real line by using `type=2` when $|z| < 1$ and `type=3` when $|z| > 1$. To fold `clpmn` and `lpmn` into a single function we've made it so that for real `z` , when `type = 2` `nan` is returned when $|z| > 1$ and for `type = 3`, `nan` is returned when $|z| < 1$. I think being able to fold these into a single function outweighs the benefits of `lpmn` being defined over the real line (with singularities at $z = \pm 1$. The two cases, `type=2` and `type=3` are different solutions of the differential equation, and my understanding is that it's not common for them to be used in the same application as if they were one function. The `type` parameter has been given the more descriptive name `branch_cut`.

New functions, `sph_legendre_p` and `sph_legendre_p_all` have also been added. `sph_legendre_p(n, m, theta)` is equal to `assoc_legendre_p(n, m, cos(theta))` and is equal to the legendre function factor in the computation of a spherical harmonic, as seen in the scipy docs [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.sph_harm.html). The spherical legendre functions are available in the [C++ standard library](https://en.cppreference.com/w/cpp/numeric/special_functions/sph_legendre) and I think are a worthy addition.

The old function `sph_harm` that we propose to deprecate used `theta` to refer to the azimuthal angle and `phi` to refer to the polar angle. This is the reverse of the convention used in most other packages, and the reverse of the most common convention for describing spherical coordinates in the literature. (This is alluded to in the documentation for [sph_harm](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.sph_harm.html).

The new function `sph_harm_y` uses `theta` for the polar angle and `phi` for the azimuthal angle. It also reverses the positional arguments to take the polar angle before the azimuthal angle, bringing it in line with the convention in the majority of other languages and packages. Here's an interesting discussion on the convention for which angles `theta` and `phi` stand for, https://www.physicsforums.com/threads/which-angle-convention-is-most-common-in-spherical-coordinates-theta-or-phi.1026197/. Since we're starting fresh with a new function, I agree with @izaid that it makes sense to bring things in line with what is more common in other languages and packages, and more in line with the most common convention for spherical coordinates in the literature.

Related work not implemented in this PR but planned for the future includes writing Legendre functions of the second kind, `assoc_legendre_q` and `legendre_q`, in the style of the functions proposed in this PR to replace `lqmn` and `lqn` respectively. Folding `lpmv` for integer order and real degree into `assoc_legendre_p`.

Internally, @izaid has come up with an elegant C++ solution for implementing special functions based on recurrences which is useful more generally than just for what has been done here. My understanding is that he has written the C++ with an eye towards making these functions available on the GPU along the lines of what was discussed in https://github.com/scipy/scipy/pull/19601. On the Python side he has implemented the `MultiUFunc` class, a pragmatic solution for creating callables work by dispatching to different ufuncs/gufuncs based on keyword arguments.

There are new, more thorough, tests for Legendre functions. The style of these tests could be improved, e.g. refactoring to use`pytest.mark.parametrize in some cases to replace having many separate asserts in the test , but I think they are sufficient for now.  I would be willing to clean them up in the future.

I've spent a considerable amount of time going through this PR, and I feel sufficiently convinced that everything is on the level from a technical perspective. I'm strongly in favor of merging this. It brings a significant improvement to `special` in my opinion and will allow @izaid to continue along his planned path to try bring these functions to the GPU through CuPy.

I propose to merge this PR this coming Tuesday if no one has any objections. I think this is a reasonable amount of time since much of what is proposed has been discussed on discourse https://discuss.scientific-python.org/t/naming-convention-for-generalized-ufuncs-in-special/1144, where @izaid pointed to this PR as an example of his proposed convention for special functions based on recurrences; and in the issue https://github.com/scipy/scipy/issues/20448. I'd be OK with extending the time period before merging if someone makes a good case that more time might be needed before we can establish a consensus.

Thanks @izaid for your hard work on this one.

Commit:
updated branch_cut docs

steppi commented on scipy/special/special/legendre.h:
```diff
@@ -1,352 +1,998 @@
+            if (std::real(z) < 0) {
+                w = -w;
             }
         } else {
```
When `branch_cut` isn't one of 2 or 3, it's just automatically treated as 2 here. I think it would be better for `assoc_legendre_p` and `assoc_legendre_p_all` to return `nan` in those cases and use `set_error` to give an informative message.

steppi replied:
I plan to do this in a follow up. The current broken behavior is lifted from `clpmn`.

steppi committed:
Fix sphinx RST issue

steppi commented:
I'm not sure why the linter is complaining about something unrelated to this PR now but didn't earlier. I assume it's fine to ignore that.

```
2271 |         # see https://github.com/numpy/numpy/issues/24050
2272 |         def assert_really_equal(x, y):
2273 |             assert type(x) == type(y), f"types not equal: {type(x)}, {type(y)}"
     |                    ^^^^^^^^^^^^^^^^^^ E721
2274 |             assert_equal(x, y)
```

steppi committed:
Mark test as slow

- Can take over 1 second to run

steppi committed:
Merge branch 'main' into legendre

lucascolley commented:
> I'm not sure why the linter is complaining about something unrelated to this PR now but didn't earlier.

New `ruff` release two days ago, so my guess is that some detection was improved.

> I assume it's fine to ignore that.

Yep, in this case `noqa: E721` looks like the correct move, as this is one of the exceptions to the rule where we don't want to accept subclasses with `isinstance`.

steppi committed:
Add noqa to ignore linter since this is desired

steppi commented:
The docs failure is unrelated, it's appearing on other PRs as well.

steppi commented:
> I propose to merge this PR this coming Tuesday if no one has any objections.

It seems that there were no objections here or on [discourse](https://discuss.scientific-python.org/t/naming-convention-for-generalized-ufuncs-in-special/1144/37?u=steppi). Thanks @izaid, let's get this in.
